### ISSUE TITLE
Get this thing deployed

### ISSUE THREAD

**Issue by lmorchard** [mozilla/watchdog-proxy#97]
We have a dev stack auto-deploying from master. We should have staging and production stacks deployed. Ideally, auto-deployed from commits to staging and production branches on this repo.

**Comment by oremj** [comment]
I was looking through https://serverless.com/framework/docs/providers/aws/guide/deploying/. We should be able to deploy using serverless, however, we can only automate `deploy function`.

As far as long lived resources dynamo, s3, etc. Those will be spun up in separate cloudformations. Reason being, there may be times when we want to keep the old version of the app running, deploy a new version and cut over without deleting or spinning up long lived resources again.

Can we change those resources to parameters? Is there any way to spin those up for development environments only?

**Comment by lmorchard** [comment]
> Can we change those resources to parameters?

Yeah. Currently they're all defined together in `serverless.yml` - but we could rework things so that queue, bucket, and table names are all passed in via environment vars and just define the functions.

> Is there any way to spin those up for development environments only?

Sure, we could split the serverless.yml into entirely separate files for prod vs dev

**Comment by lmorchard** [comment]
Still kind of tinkering with this - I thought maybe it would be as easy as supplying a command option to select a different serverless.yml. But, it turns out this option doesn't exist as such

**Comment by lmorchard** [comment]
Gave a shot at tweaking configs so that a production deploy only includes functions. [Everything else should be supplied & identified in env vars from pre-existing resources](https://github.com/lmorchard/watchdog-proxy/blob/68829464d561ab0b4e9896f1988546d3e2672bef/README.md#production-deployment). This branch / PR could use some review, though, to see if I'm on the right track.

**Comment by oremj** [comment]
For deployment stuff, you can follow along here: https://github.com/mozilla-services/cloudops-deployment/pull/2358

**Comment by oremj** [comment]
@lmorchard what are the ping centre urls?

**Comment by oremj** [comment]
https://watchdogproxy-default.stage.mozaws.net/ and https://watchdogproxy-default.dev.mozaws.net are up and running, but I'm not sure I have all the right configuration.

**Comment by lmorchard** [comment]
> what are the ping centre urls?

[Defined here in constants](https://github.com/mozilla/watchdog-proxy/blob/master/lib/constants.js#L17), but you shouldn't need to explicitly configure them since the service [should choose the production Ping Centre URL if `NODE_ENV == production`](https://github.com/mozilla/watchdog-proxy/blob/master/lib/metrics.js#L15)

**Comment by lmorchard** [comment]
> but I'm not sure I have all the right configuration

Were you able to test it? Were there any errors in CloudWatch? You should be able to run `npm run client -- --url <url>` against the endpoint if dev credentials are enabled. 

On something like stage or prod, dev credentials shouldn't be enabled (the default) so you'd need to create a row in the credentials DBD table with `id` and `key` attributes and supply those to the client with `--id <id> --key <key>`

**Comment by lmorchard** [comment]
FWIW, when I try the client against either of those endpoints, I get a `405 - "Unsupported HTTP method"` response so the API gateway is probably not set up to invoke the functions
```
C:\Users\lmorc\devel\watchdog-proxy [master ≡ +0 ~1 -0 !]> npm run client -- --url https://watchdogproxy-default.dev.mozaws.net/

> watchdog-proxy@0.0.1 client C:\Users\lmorc\devel\watchdog-proxy
> node bin/client.js "--url" "https://watchdogproxy-default.dev.mozaws.net/"

ERROR 405 - "Unsupported HTTP method"
C:\Users\lmorc\devel\watchdog-proxy [master ≡ +0 ~1 -0 !]> npm run client -- --url https://watchdogproxy-default.stage.mozaws.net/

> watchdog-proxy@0.0.1 client C:\Users\lmorc\devel\watchdog-proxy
> node bin/client.js "--url" "https://watchdogproxy-default.stage.mozaws.net/"

ERROR 405 - "Unsupported HTTP method"
```

**Comment by lmorchard** [comment]
Looks like GET requests to `__version__` and `__heartbeat__` work, though:
```
lmorchard@preciso:~$ curl -ksi https://watchdogproxy-default.dev.mozaws.net/__version__; echo                                                                                             HTTP/2 200                                                                                                                                                                                date: Fri, 10 Aug 2018 20:15:27 GMT                                                                                                                                                       content-type: application/json                                                                                                                                                            content-length: 93                                                                                                                                                                        x-amzn-requestid: 1f17b0dd-9cda-11e8-8216-91b64744aa21                                                                                                                                    x-amz-apigw-id: LbLO5G8VIAMFsdw=                                                                                                                                                          x-amzn-trace-id: Root=1-5b6df25f-4fb0ce306255e8a07b759a10;Sampled=0                                                                                                                                                                                                                                                                                                                 {"commit":"dummy","version":"0.0.1","source":"https://github.com/mozilla/watchdog-proxy.git"}             

lmorchard@preciso:~$ curl -ksi https://watchdogproxy-default.dev.mozaws.net/__version__; echo
HTTP/2 200
date: Fri, 10 Aug 2018 20:15:27 GMT
content-type: application/json
content-length: 93
x-amzn-requestid: 1f17b0dd-9cda-11e8-8216-91b64744aa21
x-amz-apigw-id: LbLO5G8VIAMFsdw=
x-amzn-trace-id: Root=1-5b6df25f-4fb0ce306255e8a07b759a10;Sampled=0

{"commit":"dummy","version":"0.0.1","source":"https://github.com/mozilla/watchdog-proxy.git"}
lmorchard@preciso:~$ curl -ksi https://watchdogproxy-default.dev.mozaws.net/__heartbeat__; echo
HTTP/2 200
date: Fri, 10 Aug 2018 20:15:54 GMT
content-type: application/json
content-length: 2
x-amzn-requestid: 2f4dc739-9cda-11e8-8cd3-1d0f7351c7cd
x-amz-apigw-id: LbLTJF-YoAMFs9Q=
x-amzn-trace-id: Root=1-5b6df27a-265dd82647cce15001902416;Sampled=0

OK
```

**Comment by lmorchard** [comment]
Also, we have a dev stack running on mozaws.net deployed via serverless. You should see something like this from the client:
```
C:\Users\lmorc\devel\watchdog-proxy [master ≡ +0 ~1 -0 !]> npm run client -- --url https://watchdog-proxy.dev.mozaws.net

> watchdog-proxy@0.0.1 client C:\Users\lmorc\devel\watchdog-proxy
> node bin/client.js "--url" "https://watchdog-proxy.dev.mozaws.net"

{"id":"b8883d5c-9cda-11e8-a6da-23f719df3a23","negative_uri":"https://watchdog-proxy.dev.mozaws.net/mock/client/negative","positive_uri":"https://watchdog-proxy.dev.mozaws.net/mock/client/positive"}
```

**Comment by oremj** [comment]
Looks like dev is working now:
```
✗ npm run client -- --url https://watchdogproxy-default.dev.mozaws.net

> watchdog-proxy@0.0.1 client /Users/oremj/Documents/work/watchdog-proxy
> node bin/client.js "--url" "https://watchdogproxy-default.dev.mozaws.net"

{"id":"b2fb4381-9cf2-11e8-8a99-bd08d9c6abf9","negative_uri":"https://watchdogproxy-default.dev.mozaws.net/mock/client/negative","positive_uri":"https://watchdogproxy-default.dev.mozaws.net/mock/client/positive"}
```

**Comment by oremj** [comment]
Stage too:
```
✗ npm run client -- --url https://watchdogproxy-default.stage.mozaws.net

> watchdog-proxy@0.0.1 client /Users/oremj/Documents/work/watchdog-proxy
> node bin/client.js "--url" "https://watchdogproxy-default.stage.mozaws.net"

{"id":"53179393-9cf3-11e8-ab6e-6f438e0f4e89","negative_uri":"https://watchdogproxy-default.stage.mozaws.net/mock/client/negative","positive_uri":"https://watchdogproxy-default.stage.mozaws.net/mock/client/positive"}
```

**Comment by oremj** [comment]
Dev should update within 10 minutes of committing to master.

**Comment by lmorchard** [comment]
Digging into this some more on the new watchdog-default stack on dev:

* Looks like the `/accept` API endpoint is working to enqueue a request.
* From there it looks like the `processQueueItem` function does get called from an SQS event.
* But, I'm seeing an error [in CloudWatch](https://console.aws.amazon.com/cloudwatch/home?region=us-east-1#logEventViewer:group=/aws/lambda/watchdog-dev-default-process-dev;stream=2018/08/13/[39]647b90d20d294bc2a149cb904006f656;start=2018-08-12T20:06:02Z) when `processQueueItem` tries to make a request out to the upstream service - `REQUEST ERROR { RequestError: Error: Invalid URI "/?enhance"`

I'm guessing that [the `UPSTREAM_SERVICE_URL` and `UPSTREAM_SERVICE_KEY` env vars](https://github.com/mozilla/watchdog-proxy#environment-variables) are not set.

In production, `UPSTREAM_SERVICE_URL` and `UPSTREAM_SERVICE_KEY` should be set to call out to PhotoDNA with the appropriate secret key. But, on dev & stage, we don't want to call out to the real service.

For that, [there are several "mock" endpoints](https://github.com/mozilla/watchdog-proxy/blob/master/config/dev.yml#L172) that I deploy to my dev stack. They're fake stand-in services for the upstream service, a positive callback, a negative callback, and metrics calls. Hits to those mock stand-ins get logged to CloudWatch so we can check the flow through the system.

I don't see any of the mock services deployed with this new dev stack, so those might need to be added for dev and stage (but not prod)

Not sure if it helps, [but there is a CloudFormation stack for my dev stack deployed via serverless that has all the functions included](https://console.aws.amazon.com/cloudformation/home?region=us-east-1#/stack/detail?stackId=arn:aws:cloudformation:us-east-1:927034868273:stack%2Fwatchdog-proxy-dev%2Fa0c49010-5d30-11e8-a7bd-50fae98b1835)

**Comment by lmorchard** [comment]
Most of this thread went to IRC, quick notes:
* Looks like the new dev stack is working with mock end points
* We moved on to a staging stack that should be sending requests to the upstream Photo DNA service. But it doesn't seem to be coming back with positive matches even when we send testing images that should cause positives.

So, I played around a bit with my personal dev stack and pointed it at the real upstream service. Hoped to find an issue here to debug. But, as soon as I got the URL correct, my stack started behaving correctly.
